`OptionsInput` event handler typings have 20+ missing properties and other errors

FullCalendar's event-triggering code is not (yet!) type safe. Not surprisingly as a result, the typings of event handlers in `OptionsInput` are missing several events and many event-argument properties that are used inside existing FullCalendar code. A few typings are simply incorrect, like `eventResize` which is typed as having a `delta` property while the triggering code in EventResizing.ts uses `startDelta` and `endDelta` but not `delta`. 

These inaccurate typings make it difficult to accurately write event handlers in client apps because important properties are missing or wrong. This leads to compiler errors, runtime errors, or both.

To repro: 
```
git clone https://github.com/justingrant/fullcalendar.git
cd fullcalendar
git checkout trigger-type-errors
npm install
npm run-script tsc
```

The repo above is identical to the current fullcalendar/master branch, except that it adds type checking to event triggers so that any time a triggered event doesn't match the typings, a compiler error will be generated. (I'm working on a PR that includes this code and which also fixes any typings errors exposed)

Missing or inaccurate typings include: 
* `eventLimitClick` - the `cellInfo` property is listed in typings but isn't used in FullCalendar code, and the event type is missing the following properties: `date`, `allDay`, `dayEl`, `moreEl`, `segs`, `hiddenSegs`, `view`. Also the static `'timeGridWeek'` and `'timeGridDay'` options are also missing. 
* `dayRender` - the `allDay` property should be optional because it's not always used when the event is triggered
* `dateClick` - the `resource` property should be optional because it's not always used when the event is triggered
* `select` the `resource` property should be optional because it's not always used when the event is triggered
* `loading` - this event is typed as having a second parameter named `view`, but this parameter is never included anywhere this event is triggered.
* `eventRender` - the `isMirror`, `isStart`, and `isEnd` properties are missing
* `eventPositioned` - the `isMirror`, `isStart`, and `isEnd` properties are missing
* `eventDestroy` - the `isMirror` property is missing
* `eventDrop` - the `oldEvent` property is missing
* `eventResize` - a `delta` property is included in the current typings, but the actual properties triggered are `startDelta` and `endDelta`. In addition, the `prevEvent` property is missing from typings. 
* The following events are missing from typings altogether: ` _init`, `_noEventDrop`, and `_noEventResize`

Included below is compiler output you'll see when compiling the repro repo.  Here's a key to the problems you'll see: 
* The `not assignable to parameter of type 'EventHandlerName'` errors denote events that are missing in current `OptionsInput` declaration.
* The `Property 'XXXX' is missing` errors are missing properties in an event arg declaration.
* The `not assignable to type` errors are where the triggering code is passing an event argument that doesn't match the expected type as declared in the TS declaration. Usually this is because the triggering code is missing a cast (e.g. UIEvent->MouseEvent) but some cases look more serious, like #4644.

```
src/core/Calendar.ts:135:26 - error TS2345: Argument of type '"_init"' is not assignable to parameter of type 'EventHandlerName'.

135     this.publiclyTrigger('_init') // for tests
                             ~~~~~~~

src/core/Calendar.ts:307:41 - error TS2345: Argument of type '[true]' is not assignable to parameter of type '[boolean, View]'.
  Property '1' is missing in type '[true]' but required in type '[boolean, View]'.

307         this.publiclyTrigger('loading', [ true ])
                                            ~~~~~~~~

src/core/Calendar.ts:309:41 - error TS2345: Argument of type '[false]' is not assignable to parameter of type '[boolean, View]'.
  Property '1' is missing in type '[false]' but required in type '[boolean, View]'.

309         this.publiclyTrigger('loading', [ false ])
                                            ~~~~~~~~~

src/core/Calendar.ts:1061:38 - error TS2741: Property 'resource' is missing in type 'DateSelectionApi' but required in type '{ start: Date; end: Date; startStr: string; endStr: string; allDay: boolean; resource: any; jsEvent: MouseEvent; view: View; }'.

1061     this.publiclyTrigger('select', [ arg ])
                                          ~~~

  src/core/types/input-types.ts:201:93
    201   select?(arg: { start: Date, end: Date, startStr: string, endStr: string, allDay: boolean, resource: any, jsEvent: MouseEvent, view: View }): void // resource for Scheduler
                                                                                                    ~~~~~~~~
    'resource' is declared here.

src/core/Calendar.ts:1083:41 - error TS2741: Property 'resource' is missing in type 'DateClickApi' but required in type '{ date: Date; dateStr: string; allDay: boolean; resource: any; dayEl: HTMLElement; jsEvent: MouseEvent; view: View; }'.

1083     this.publiclyTrigger('dateClick', [ arg ])
                                             ~~~

  src/core/types/input-types.ts:197:67
    197   dateClick?(arg: { date: Date, dateStr: string, allDay: boolean, resource: any, dayEl: HTMLElement, jsEvent: MouseEvent, view: View }): void // resource for Scheduler
                                                                          ~~~~~~~~
    'resource' is declared here.

src/core/component/DateComponent.ts:201:13 - error TS2322: Type 'this' is not assignable to type 'View'.
  Type 'DateComponent<PropsType>' is missing the following properties from type 'View': usesMinMaxTime, dateProfileGeneratorClass, on, one, and 70 more.

201             view: this // ?
                ~~~~

  src/core/types/input-types.ts:205:61
    205   eventPositioned?(arg: { event: EventApi, el: HTMLElement, view: View }): void
                                                                    ~~~~
    The expected type comes from property 'view' which is declared here on type '{ event: EventApi; el: HTMLElement; view: View; }'

src/core/component/DateComponent.ts:231:13 - error TS2322: Type 'this' is not assignable to type 'View'.

231             view: this // ?
                ~~~~

  src/core/types/input-types.ts:207:58
    207   eventDestroy?(arg: { event: EventApi, el: HTMLElement, view: View }): void
                                                                 ~~~~
    The expected type comes from property 'view' which is declared here on type '{ event: EventApi; el: HTMLElement; view: View; }'

src/core/component/event-rendering.ts:131:11 - error TS2322: Type '{ event: EventApi; isMirror: any; isStart: boolean; isEnd: boolean; el: HTMLElement; view: View; }' is not assignable to type '{ event: EventApi; el: HTMLElement; view: View; }'.
  Object literal may only specify known properties, and 'isMirror' does not exist in type '{ event: EventApi; el: HTMLElement; view: View; }'.

131           isMirror,
              ~~~~~~~~

src/core/interactions/EventClicking.ts:46:11 - error TS2740: Type 'Event' is missing the following properties from type 'MouseEvent': altKey, button, buttons, clientX, and 25 more.

46           jsEvent: ev,
             ~~~~~~~

  src/core/types/input-types.ts:198:56
    198   eventClick?(arg: { el: HTMLElement, event: EventApi, jsEvent: MouseEvent, view: View }): boolean | void
                                                               ~~~~~~~
    The expected type comes from property 'jsEvent' which is declared here on type '{ el: HTMLElement; event: EventApi; jsEvent: MouseEvent; view: View; }'

src/core/interactions/EventHovering.ts:62:33 - error TS2345: Argument of type 'string' is not assignable to parameter of type 'EventHandlerName'.

62       component.publiclyTrigger(publicEvName, [
                                   ~~~~~~~~~~~~

src/daygrid/DayGrid.ts:250:11 - error TS2741: Property 'allDay' is missing in type '{ date: Date; el: HTMLElement; view: View; }' but required in type '{ view: View; date: Date; allDay: boolean; el: HTMLElement; }'.

250           {
              ~
251             date: dateEnv.toDate(cells[row][col].date),
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
... 
253             view
    ~~~~~~~~~~~~~~~~
254           }
    ~~~~~~~~~~~

  src/core/types/input-types.ts:195:45
    195   dayRender?(arg: { view: View, date: Date, allDay: boolean, el: HTMLElement }): void
                                                    ~~~~~~
    'allDay' is declared here.

src/daygrid/DayGrid.ts:731:13 - error TS2322: Type '{ date: Date; allDay: boolean; dayEl: HTMLElement; moreEl: HTMLElement; segs: any[]; hiddenSegs: any[]; jsEvent: MouseEvent; view: View; }' is not assignable to type 'CellInfo'.
  Object literal may only specify known properties, and 'allDay' does not exist in type 'CellInfo'.

731             allDay: true,
                ~~~~~~~~~~~~

src/interaction/interactions-external/ExternalElementDragging.ts:139:51 - error TS2322: Type 'ExternalDropApi' is not assignable to type '{ date: Date; dateStr: string; allDay: boolean; draggedEl: HTMLElement; jsEvent: MouseEvent; view: View; }'.

139       receivingCalendar.publiclyTrigger('drop', [ arg ])
                                                      ~~~

src/interaction/interactions-external/ExternalElementDragging.ts:157:13 - error TS2322: Type 'EventTarget' is not assignable to type 'HTMLElement'.

157             draggedEl: pev.subjectEl,
                ~~~~~~~~~

  src/core/types/input-types.ts:215:41
    215   eventReceive?(arg: { event: EventApi, draggedEl: HTMLElement, view: View }): void
                                                ~~~~~~~~~
    The expected type comes from property 'draggedEl' which is declared here on type '{ event: EventApi; draggedEl: HTMLElement; view: View; }'

src/interaction/interactions/EventDragging.ts:118:11 - error TS2322: Type 'UIEvent' is not assignable to type 'MouseEvent'.

118           jsEvent: ev.origEvent,
              ~~~~~~~

  src/core/types/input-types.ts:208:60
    208   eventDragStart?(arg: { event: EventApi, el: HTMLElement, jsEvent: MouseEvent, view: View }): void
                                                                   ~~~~~~~
    The expected type comes from property 'jsEvent' which is declared here on type '{ event: EventApi; el: HTMLElement; jsEvent: MouseEvent; view: View; }'

src/interaction/interactions/EventDragging.ts:231:11 - error TS2322: Type 'UIEvent' is not assignable to type 'MouseEvent'.

231           jsEvent: ev.origEvent,
              ~~~~~~~

  src/core/types/input-types.ts:209:59
    209   eventDragStop?(arg: { event: EventApi, el: HTMLElement, jsEvent: MouseEvent, view: View }): void
                                                                  ~~~~~~~
    The expected type comes from property 'jsEvent' which is declared here on type '{ event: EventApi; el: HTMLElement; jsEvent: MouseEvent; view: View; }'

src/interaction/interactions/EventDragging.ts:271:58 - error TS2740: Type '{}' is missing the following properties from type '{ el: HTMLElement; event: EventApi; delta: Duration; revert: () => void; jsEvent: Event; view: View; }': el, event, delta, revert, and 2 more.

271           initialCalendar.publiclyTrigger('eventDrop', [ eventDropArg ])
                                                             ~~~~~~~~~~~~

src/interaction/interactions/EventDragging.ts:278:15 - error TS2740: Type 'EventTarget' is missing the following properties from type 'HTMLElement': accessKey, accessKeyLabel, autocapitalize, dir, and 234 more.

278               draggedEl: ev.subjectEl,
                  ~~~~~~~~~

  src/core/types/input-types.ts:216:22
    216   eventLeave?(arg: { draggedEl: HTMLElement, event: EventApi, view: View }): void
                             ~~~~~~~~~
    The expected type comes from property 'draggedEl' which is declared here on type '{ draggedEl: HTMLElement; event: EventApi; view: View; }'

src/interaction/interactions/EventDragging.ts:306:55 - error TS2322: Type 'ExternalDropApi' is not assignable to type '{ date: Date; dateStr: string; allDay: boolean; draggedEl: HTMLElement; jsEvent: MouseEvent; view: View; }'.
  Types of property 'jsEvent' are incompatible.
    Type 'UIEvent' is not assignable to type 'MouseEvent'.

306           receivingCalendar.publiclyTrigger('drop', [ dropArg ])
                                                          ~~~~~~~

src/interaction/interactions/EventDragging.ts:310:15 - error TS2322: Type 'EventTarget' is not assignable to type 'HTMLElement'.

310               draggedEl: ev.subjectEl,
                  ~~~~~~~~~

  src/core/types/input-types.ts:215:41
    215   eventReceive?(arg: { event: EventApi, draggedEl: HTMLElement, view: View }): void
                                                ~~~~~~~~~
    The expected type comes from property 'draggedEl' which is declared here on type '{ event: EventApi; draggedEl: HTMLElement; view: View; }'

src/interaction/interactions/EventDragging.ts:316:15 - error TS2740: Type 'DateComponent<any>' is missing the following properties from type 'View': usesMinMaxTime, dateProfileGeneratorClass, on, one, and 70 more.

316               view: finalHit.component
                  ~~~~

  src/core/types/input-types.ts:215:65
    215   eventReceive?(arg: { event: EventApi, draggedEl: HTMLElement, view: View }): void
                                                                        ~~~~
    The expected type comes from property 'view' which is declared here on type '{ event: EventApi; draggedEl: HTMLElement; view: View; }'

src/interaction/interactions/EventDragging.ts:322:41 - error TS2345: Argument of type '"_noEventDrop"' is not assignable to parameter of type 'EventHandlerName'.

322         initialCalendar.publiclyTrigger('_noEventDrop')
                                            ~~~~~~~~~~~~~~

src/interaction/interactions/EventResizing.ts:83:9 - error TS2322: Type 'UIEvent' is not assignable to type 'MouseEvent'.

83         jsEvent: ev.origEvent,
           ~~~~~~~

  src/core/types/input-types.ts:211:62
    211   eventResizeStart?(arg: { el: HTMLElement, event: EventApi, jsEvent: MouseEvent, view: View }): void
                                                                     ~~~~~~~
    The expected type comes from property 'jsEvent' which is declared here on type '{ el: HTMLElement; event: EventApi; jsEvent: MouseEvent; view: View; }'

src/interaction/interactions/EventResizing.ts:166:9 - error TS2740: Type 'UIEvent' is missing the following properties from type 'MouseEvent': altKey, button, buttons, clientX, and 22 more.

166         jsEvent: ev.origEvent,
            ~~~~~~~

  src/core/types/input-types.ts:212:61
    212   eventResizeStop?(arg: { el: HTMLElement, event: EventApi, jsEvent: MouseEvent, view: View }): void
                                                                    ~~~~~~~
    The expected type comes from property 'jsEvent' which is declared here on type '{ el: HTMLElement; event: EventApi; jsEvent: MouseEvent; view: View; }'

src/interaction/interactions/EventResizing.ts:180:11 - error TS2322: Type '{ el: HTMLElement; startDelta: Duration; endDelta: Duration; prevEvent: EventApi; event: EventApi; revert: () => void; jsEvent: UIEvent; view: View; }' is not assignable to type '{ el: HTMLElement; event: EventApi; delta: Duration; revert: () => void; jsEvent: Event; view: View; }'.
  Object literal may only specify known properties, and 'startDelta' does not exist in type '{ el: HTMLElement; event: EventApi; delta: Duration; revert: () => void; jsEvent: Event; view: View; }'.

180           startDelta: this.validMutation.startDelta || createDuration(0),
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/interaction/interactions/EventResizing.ts:200:32 - error TS2345: Argument of type '"_noEventResize"' is not assignable to parameter of type 'EventHandlerName'.

200       calendar.publiclyTrigger('_noEventResize')
                                   ~~~~~~~~~~~~~~~~

src/timegrid/TimeGrid.ts:386:9 - error TS2741: Property 'allDay' is missing in type '{ date: Date; el: HTMLElement; view: View; }' but required in type '{ view: View; date: Date; allDay: boolean; el: HTMLElement; }'.

386         {
            ~
387           date: dateEnv.toDate(cells[col].date),
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
... 
389           view
    ~~~~~~~~~~~~~~
390         }
    ~~~~~~~~~

  src/core/types/input-types.ts:195:45
    195   dayRender?(arg: { view: View, date: Date, allDay: boolean, el: HTMLElement }): void
                                                    ~~~~~~
    'allDay' is declared here.
```

--- a/src/core/Calendar.ts
+++ b/src/core/Calendar.ts
@@ -4,7 +4,7 @@ import { default as EmitterMixin, EmitterInterface } from './common/EmitterMixin
 import OptionsManager from './OptionsManager'
 import View from './View'
 import Theme from './theme/Theme'
-import { OptionsInput } from './types/input-types'
+import { OptionsInput, EventHandlerName, EventHandlerArgs } from './types/input-types'
 import { Locale, buildLocale, parseRawLocales, RawLocaleMap } from './datelib/locale'
 import { DateEnv, DateInput } from './datelib/env'
 import { DateMarker, startOfDay } from './datelib/marker'
@@ -702,13 +702,13 @@ export default class Calendar {
   // -----------------------------------------------------------------------------------------------------------------
 
 
-  hasPublicHandlers(name: string): boolean {
+  hasPublicHandlers<T extends EventHandlerName>(name: T): boolean {
     return this.hasHandlers(name) ||
       this.opt(name) // handler specified in options
   }
 
 
-  publiclyTrigger(name: string, args?) {
+  publiclyTrigger<T extends EventHandlerName>(name: T, args?: EventHandlerArgs<T>) {
     let optHandler = this.opt(name)
 
     this.triggerWith(name, this, args)
@@ -719,7 +719,7 @@ export default class Calendar {
   }
 
 
-  publiclyTriggerAfterSizing(name, args) {
+  publiclyTriggerAfterSizing<T extends EventHandlerName>(name: T, args: EventHandlerArgs<T>) {
     let { afterSizingTriggers } = this;
 
     (afterSizingTriggers[name] || (afterSizingTriggers[name] = [])).push(args)
@@ -731,7 +731,7 @@ export default class Calendar {
 
     for (let name in afterSizingTriggers) {
       for (let args of afterSizingTriggers[name]) {
-        this.publiclyTrigger(name, args)
+        this.publiclyTrigger(name as EventHandlerName, args)
       }
     }
 
@@ -1048,11 +1048,11 @@ export default class Calendar {
 
 
   triggerDateSelect(selection: DateSpan, pev?: PointerDragEvent) {
-    let arg = this.buildDateSpanApi(selection) as DateSelectionApi
-
-    arg.jsEvent = pev ? pev.origEvent : null
-    arg.view = this.view
-
+    const arg = {
+      ...this.buildDateSpanApi(selection),
+      jsEvent: pev ? pev.origEvent as MouseEvent : null, // Is this always a mouse event? See #4655
+      view: this.view
+    }
     this.publiclyTrigger('select', [ arg ])
   }
 
@@ -1069,11 +1069,12 @@ export default class Calendar {
 
   // TODO: receive pev?
   triggerDateClick(dateSpan: DateSpan, dayEl: HTMLElement, view: View, ev: UIEvent) {
-    let arg = this.buildDatePointApi(dateSpan) as DateClickApi
-
-    arg.dayEl = dayEl
-    arg.jsEvent = ev
-    arg.view = view
+    const arg = {
+      ...this.buildDatePointApi(dateSpan),
+      dayEl,
+      jsEvent: ev as MouseEvent, // Is this always a mouse event? See #4655
+      view
+    }
 
     this.publiclyTrigger('dateClick', [ arg ])
   }

--- a/src/core/component/DateComponent.ts
+++ b/src/core/component/DateComponent.ts
@@ -10,6 +10,8 @@ import EventApi from '../api/EventApi'
 import FgEventRenderer from './renderers/FgEventRenderer'
 import FillRenderer from './renderers/FillRenderer'
 import { EventInteractionState } from '../interactions/event-interaction-state'
+import View from '../View'
+import { EventHandlerName, EventHandlerArgs } from '../types/input-types'
 
 export type DateComponentHash = { [uid: string]: DateComponent<any> }
 
@@ -153,21 +155,21 @@ export default class DateComponent<PropsType> extends Component<PropsType> {
   // TODO: move to Calendar
 
 
-  publiclyTrigger(name, args) {
+  publiclyTrigger<T extends EventHandlerName>(name: T, args?: EventHandlerArgs<T>) {
     let calendar = this.calendar
 
     return calendar.publiclyTrigger(name, args)
   }
 
 
-  publiclyTriggerAfterSizing(name, args) {
+  publiclyTriggerAfterSizing<T extends EventHandlerName>(name: T, args: EventHandlerArgs<T>) {
     let calendar = this.calendar
 
     return calendar.publiclyTriggerAfterSizing(name, args)
   }
 
 
-  hasPublicHandlers(name) {
+  hasPublicHandlers<T extends EventHandlerName>(name: T) {
     let calendar = this.calendar
 
     return calendar.hasPublicHandlers(name)
@@ -191,7 +193,7 @@ export default class DateComponent<PropsType> extends Component<PropsType> {
             isStart: seg.isStart,
             isEnd: seg.isEnd,
             el: seg.el,
-            view: this // ?
+            view: this as unknown as View // safe to cast because this method is only called on context.view
           }
         ])
       }
@@ -221,7 +223,7 @@ export default class DateComponent<PropsType> extends Component<PropsType> {
             ),
             isMirror: isMirrors,
             el: seg.el,
-            view: this // ?
+            view: this as unknown as View // safe to cast because this method is only called on context.view
           }
         ])
       }

--- a/src/core/component/event-rendering.ts
+++ b/src/core/component/event-rendering.ts
@@ -117,7 +117,7 @@ export function hasBgRendering(def: EventDef) {
   return def.rendering === 'background' || def.rendering === 'inverse-background'
 }
 
-export function filterSegsViaEls(view: View, segs: Seg[], isMirror) {
+export function filterSegsViaEls(view: View, segs: Seg[], isMirror: boolean) {
 
   if (view.hasPublicHandlers('eventRender')) {
     segs = segs.filter(function(seg) {

--- a/src/core/interactions/EventClicking.ts
+++ b/src/core/interactions/EventClicking.ts
@@ -43,7 +43,7 @@ export default class EventClicking extends Interaction {
             seg.eventRange.def,
             seg.eventRange.instance
           ),
-          jsEvent: ev,
+          jsEvent: ev as MouseEvent, // Is this always a mouse event? See #4655
           view: component.view
         }
       ])

--- a/src/core/interactions/EventHovering.ts
+++ b/src/core/interactions/EventHovering.ts
@@ -54,7 +54,7 @@ export default class EventHovering extends Interaction {
     }
   }
 
-  triggerEvent(publicEvName: string, ev: Event | null, segEl: HTMLElement) {
+  triggerEvent(publicEvName: 'eventMouseEnter' | 'eventMouseLeave', ev: Event | null, segEl: HTMLElement) {
     let { component } = this
     let seg = getElSeg(segEl)!
 
@@ -67,7 +67,7 @@ export default class EventHovering extends Interaction {
             seg.eventRange.def,
             seg.eventRange.instance
           ),
-          jsEvent: ev,
+          jsEvent: ev as MouseEvent, // Is this always a mouse event? See #4655
           view: component.view
         }
       ])

--- a/src/core/interactions/event-dragging.ts
+++ b/src/core/interactions/event-dragging.ts
@@ -1,6 +1,7 @@
 import Calendar from '../Calendar'
 import { EventMutation } from '../structs/event-mutation'
 import { Hit } from './hit'
+import { EventHandlerArg } from '../types/input-types'
 
 export type eventDragMutationMassager = (mutation: EventMutation, hit0: Hit, hit1: Hit) => void
-export type EventDropTransformers = (mutation: EventMutation, calendar: Calendar) => void
+export type EventDropTransformers = (mutation: EventMutation, calendar: Calendar) => Partial<EventHandlerArg<'eventDrop'>>

--- a/src/core/types/input-types.ts
+++ b/src/core/types/input-types.ts
@@ -70,6 +70,23 @@ export interface DropInfo {
   end: Date
 }
 
+// TODO: refactor OptionsInputBase to split out event handlers into a separate interface,
+// which will enable replacing the static list of event handlers below with a simpler
+// `keyof OptionsInputBaseEventHandlers`
+export type EventHandlerName =
+  '_init' | 'selectAllow' | 'eventAllow' | 'eventDataTransform' | 'datesRender' |
+  'datesDestroy' | 'dayRender' | 'windowResize' | 'dateClick' | 'eventClick' |
+  'eventMouseEnter' | 'eventMouseLeave' | 'select' | 'unselect' | 'loading' |
+  'eventRender' | 'eventPositioned' | '_eventsPositioned' | 'eventDestroy' |
+  'eventDragStart' | 'eventDragStop' | 'eventDrop' | '_destroyed' | 'drop' |
+  'eventResizeStart' | 'eventResizeStop' | 'eventResize' | 'eventReceive' |
+  'eventLeave' | 'viewSkeletonRender' | 'viewSkeletonDestroy' | '_noEventDrop' |
+  '_noEventResize' | 'eventLimitClick'
+
+export type EventHandlerArgs<T extends EventHandlerName> =
+  Parameters<Extract<OptionsInput[T], (...args: any[]) => any>>
+export type EventHandlerArg<T extends EventHandlerName> = EventHandlerArgs<T>[0]
+
 export interface OptionsInputBase {
   header?: boolean | ToolbarInput
   footer?: boolean | ToolbarInput
@@ -93,7 +110,8 @@ export interface OptionsInputBase {
   handleWindowResize?: boolean
   windowResizeDelay?: number
   eventLimit?: boolean | number
-  eventLimitClick?: 'popover' | 'week' | 'day' | string | ((cellinfo: CellInfo, jsevent: Event) => void)
+  eventLimitClick?: 'popover' | 'week' | 'day' | 'timeGridWeek' | 'timeGridDay' | string |
+    ((arg: { date: Date, allDay: boolean, dayEl: HTMLElement, moreEl: HTMLElement, segs: any[], hiddenSegs: any[], jsEvent: MouseEvent, view: View }) => void),
   timeZone?: string | boolean
   now?: DateInput | (() => DateInput)
   defaultView?: string
@@ -178,31 +196,34 @@ export interface OptionsInputBase {
   timeGridEventMinHeight?: number
   datesRender?(arg: { view: View, el: HTMLElement }): void
   datesDestroy?(arg: { view: View, el: HTMLElement }): void
-  dayRender?(arg: { view: View, date: Date, allDay: boolean, el: HTMLElement }): void
+  dayRender?(arg: { view: View, date: Date, allDay?: boolean, el: HTMLElement }): void
   windowResize?(view: View): void
-  dateClick?(arg: { date: Date, dateStr: string, allDay: boolean, resource: any, dayEl: HTMLElement, jsEvent: MouseEvent, view: View }): void // resource for Scheduler
+  dateClick?(arg: { date: Date, dateStr: string, allDay: boolean, resource?: any, dayEl: HTMLElement, jsEvent: MouseEvent, view: View }): void // resource for Scheduler
   eventClick?(arg: { el: HTMLElement, event: EventApi, jsEvent: MouseEvent, view: View }): boolean | void
   eventMouseEnter?(arg: { el: HTMLElement, event: EventApi, jsEvent: MouseEvent, view: View }): void
   eventMouseLeave?(arg: { el: HTMLElement, event: EventApi, jsEvent: MouseEvent, view: View }): void
-  select?(arg: { start: Date, end: Date, startStr: string, endStr: string, allDay: boolean, resource: any, jsEvent: MouseEvent, view: View }): void // resource for Scheduler
+  select?(arg: { start: Date, end: Date, startStr: string, endStr: string, allDay: boolean, resource?: any, jsEvent: MouseEvent, view: View }): void // resource for Scheduler
   unselect?(arg: { view: View, jsEvent: Event }): void
-  loading?(isLoading: boolean, view: View): void
-  eventRender?(arg: { event: EventApi, el: HTMLElement, view: View }): void
-  eventPositioned?(arg: { event: EventApi, el: HTMLElement, view: View }): void
+  loading?(isLoading: boolean): void
+  eventRender?(arg: { isMirror: boolean, isStart: boolean, isEnd: boolean, event: EventApi, el: HTMLElement, view: View }): void
+  eventPositioned?(arg: { isMirror: boolean, isStart: boolean, isEnd: boolean, event: EventApi, el: HTMLElement, view: View }): void
   _eventsPositioned?(arg: { view: View }): void
-  eventDestroy?(arg: { event: EventApi, el: HTMLElement, view: View }): void
+  eventDestroy?(arg: { isMirror: boolean, event: EventApi, el: HTMLElement, view: View }): void
   eventDragStart?(arg: { event: EventApi, el: HTMLElement, jsEvent: MouseEvent, view: View }): void
   eventDragStop?(arg: { event: EventApi, el: HTMLElement, jsEvent: MouseEvent, view: View }): void
-  eventDrop?(arg: { el: HTMLElement, event: EventApi, delta: Duration, revert: () => void, jsEvent: Event, view: View }): void
+  eventDrop?(arg: { el: HTMLElement, event: EventApi, oldEvent: EventApi, delta: Duration, revert: () => void, jsEvent: Event, view: View }): void
   eventResizeStart?(arg: { el: HTMLElement, event: EventApi, jsEvent: MouseEvent, view: View }): void
   eventResizeStop?(arg: { el: HTMLElement, event: EventApi, jsEvent: MouseEvent, view: View }): void
-  eventResize?(arg: { el: HTMLElement, event: EventApi, delta: Duration, revert: () => void, jsEvent: Event, view: View }): void
+  eventResize?(arg: { el: HTMLElement, startDelta: Duration, endDelta: Duration, prevEvent: EventApi, event: EventApi, revert: () => void, jsEvent: Event, view: View }): void
   drop?(arg: { date: Date, dateStr: string, allDay: boolean, draggedEl: HTMLElement, jsEvent: MouseEvent, view: View }): void
   eventReceive?(arg: { event: EventApi, draggedEl: HTMLElement, view: View }): void
   eventLeave?(arg: { draggedEl: HTMLElement, event: EventApi, view: View }): void
   viewSkeletonRender?(arg: { el: HTMLElement, view: View }): void
   viewSkeletonDestroy?(arg: { el: HTMLElement, view: View }): void
   _destroyed?(): void
+  _init?(): void
+  _noEventDrop?(): void
+  _noEventResize?(): void
 }
 
 export interface ViewOptionsInput extends OptionsInputBase {

--- a/src/interaction/interactions-external/ExternalElementDragging.ts
+++ b/src/interaction/interactions-external/ExternalElementDragging.ts
@@ -130,12 +130,12 @@ export default class ExternalElementDragging {
       let finalHit = this.hitDragging.finalHit!
       let finalView = finalHit.component.view
       let dragMeta = this.dragMeta!
-      let arg = receivingCalendar.buildDatePointApi(finalHit.dateSpan) as ExternalDropApi
-
-      arg.draggedEl = pev.subjectEl as HTMLElement
-      arg.jsEvent = pev.origEvent
-      arg.view = finalView
-
+      let arg = {
+        ...receivingCalendar.buildDatePointApi(finalHit.dateSpan),
+        draggedEl: pev.subjectEl as HTMLElement,
+        jsEvent: pev.origEvent as MouseEvent, // Is this always a mouse event? See #4655
+        view: finalView
+      }
       receivingCalendar.publiclyTrigger('drop', [ arg ])
 
       if (dragMeta.create) {
@@ -154,7 +154,7 @@ export default class ExternalElementDragging {
         // signal that an external event landed
         receivingCalendar.publiclyTrigger('eventReceive', [
           {
-            draggedEl: pev.subjectEl,
+            draggedEl: pev.subjectEl as HTMLElement,
             event: new EventApi(
               receivingCalendar,
               droppableEvent.def,

--- a/src/interaction/interactions/EventDragging.ts
+++ b/src/interaction/interactions/EventDragging.ts
@@ -17,7 +17,7 @@ import {
 import HitDragging, { isHitsEqual } from './HitDragging'
 import FeaturefulElementDragging from '../dnd/FeaturefulElementDragging'
 import { __assign } from 'tslib'
-import { ExternalDropApi } from '../interactions-external/ExternalElementDragging'
+import { EventDropTransformers } from '../../core/interactions/event-dragging'
 
 
 export default class EventDragging extends Interaction { // TODO: rename to EventSelectingAndDragging
@@ -115,7 +115,7 @@ export default class EventDragging extends Interaction { // TODO: rename to Even
         {
           el: this.subjectSeg.el,
           event: new EventApi(initialCalendar, eventRange.def, eventRange.instance),
-          jsEvent: ev.origEvent,
+          jsEvent: ev.origEvent as MouseEvent, // Is this always a mouse event? See #4655
           view: this.component.view
         }
       ])
@@ -228,7 +228,7 @@ export default class EventDragging extends Interaction { // TODO: rename to Even
         {
           el: this.subjectSeg.el,
           event: eventApi,
-          jsEvent: ev.origEvent,
+          jsEvent: ev.origEvent as MouseEvent, // Is this always a mouse event? See #4655
           view: initialView
         }
       ])
@@ -243,14 +243,15 @@ export default class EventDragging extends Interaction { // TODO: rename to Even
             eventStore: mutatedRelevantEvents
           })
 
-          let eventDropArg = {}
+          let transformed: ReturnType<EventDropTransformers> = {}
 
           for (let transformer of initialCalendar.pluginSystem.hooks.eventDropTransformers) {
-            __assign(eventDropArg, transformer(this.validMutation, initialCalendar))
+            __assign(transformed, transformer(this.validMutation, initialCalendar))
           }
 
-          __assign(eventDropArg, {
-            el: ev.subjectEl,
+          const eventDropArg = {
+            ...transformed, // don't use __assign here because it's not type-safe
+            el: ev.subjectEl as HTMLElement,
             delta: this.validMutation.startDelta!,
             oldEvent: eventApi,
             event: new EventApi( // the data AFTER the mutation
@@ -266,7 +267,7 @@ export default class EventDragging extends Interaction { // TODO: rename to Even
             },
             jsEvent: ev.origEvent,
             view: initialView
-          })
+          }
 
           initialCalendar.publiclyTrigger('eventDrop', [ eventDropArg ])
 
@@ -275,7 +276,7 @@ export default class EventDragging extends Interaction { // TODO: rename to Even
 
           initialCalendar.publiclyTrigger('eventLeave', [
             {
-              draggedEl: ev.subjectEl,
+              draggedEl: ev.subjectEl as HTMLElement,
               event: eventApi,
               view: initialView
             }
@@ -298,22 +299,23 @@ export default class EventDragging extends Interaction { // TODO: rename to Even
             })
           }
 
-          let dropArg = receivingCalendar.buildDatePointApi(finalHit.dateSpan) as ExternalDropApi
-          dropArg.draggedEl = ev.subjectEl as HTMLElement
-          dropArg.jsEvent = ev.origEvent
-          dropArg.view = finalHit.component as View // ?
-
+          let dropArg = {
+            ...receivingCalendar.buildDatePointApi(finalHit.dateSpan),
+            draggedEl: ev.subjectEl as HTMLElement,
+            jsEvent: ev.origEvent as MouseEvent, // Is this always a mouse event? See #4655
+            view: finalHit.component as View // should this be finalHit.component.view? See #4644
+          }
           receivingCalendar.publiclyTrigger('drop', [ dropArg ])
 
           receivingCalendar.publiclyTrigger('eventReceive', [
             {
-              draggedEl: ev.subjectEl,
+              draggedEl: ev.subjectEl as HTMLElement,
               event: new EventApi( // the data AFTER the mutation
                 receivingCalendar,
                 mutatedRelevantEvents.defs[eventDef.defId],
                 mutatedRelevantEvents.instances[eventInstance.instanceId]
               ),
-              view: finalHit.component
+              view: finalHit.component as View // should this be finalHit.component.view? See #4644
             }
           ])
         }

--- a/src/interaction/interactions/EventResizing.ts
+++ b/src/interaction/interactions/EventResizing.ts
@@ -80,7 +80,7 @@ export default class EventDragging extends Interaction {
       {
         el: this.draggingSeg.el,
         event: new EventApi(calendar, eventRange.def, eventRange.instance),
-        jsEvent: ev.origEvent,
+        jsEvent: ev.origEvent as MouseEvent, // Is this always a mouse event? See #4655
         view: this.component.view
       }
     ])
@@ -163,7 +163,7 @@ export default class EventDragging extends Interaction {
       {
         el: this.draggingSeg.el,
         event: eventApi,
-        jsEvent: ev.origEvent,
+        jsEvent: ev.origEvent as MouseEvent, // Is this always a mouse event? See #4655
         view
       }
     ])

